staging/src/k8s.io/apiserver/pkg/endpoints/handlers/watch.go

```diff
@@ -360,3 +372,38 @@ func (s *WatchServer) HandleWS(ws *websocket.Conn) {
 		}
 	}
 }
+
+func recordWatchListLatency(ctx context.Context, reqScope *RequestScope) {
+	requestReceivedTimestamp, ok := request.ReceivedTimestampFrom(ctx)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unable to measure watchlist latency because no received ts found in the ctx, gvr: %s", reqScope.Resource))
+		return
+	}
+	requestInfo, ok := request.RequestInfoFrom(ctx)
+	if !ok {
+		utilruntime.HandleError(fmt.Errorf("unable to measure watchlist latency because no req info found in the ctx, gvr: %s", reqScope.Resource))
+		return
+	}
+	// TODO(p0lyn0mial): figure out how to get ns and name for watch requests
+	scope := metrics.CleanScope(requestInfo)
+	elapsedSeconds := time.Now().Sub(requestReceivedTimestamp).Seconds()
+	gvr := reqScope.Resource
+
+	metrics.WatchListLatencies.WithContext(ctx).WithLabelValues(gvr.Group, gvr.Version, gvr.Resource, scope).Observe(elapsedSeconds)
+}
+
+func shouldRecordWatchListLatency(event watch.Event) bool {
+	if event.Type != watch.Bookmark || !utilfeature.DefaultFeatureGate.Enabled(features.WatchList) {
+		return false
+	}
+	// as of today the initial-events-end annotation is added only to a single event
+	// by the watch cache and only when certain conditions are met
+	//
+	// for more please read https://github.com/kubernetes/enhancements/tree/master/keps/sig-api-machinery/3157-watch-list
+	hasAnnotation, err := storage.HasInitialEventsEndBookmarkAnnotation(event.Object)
```

Comment:
Is my understand of the KEP correct, that this means that we will only record watch requests that succeeded to do the initial sync and not include the ones from slow informers that might've taken more than 60 seconds and timed out?

Is this a scenario that is intentionally excluded from the measurements?

Reply:
In general watch request have longer timeout than 60 seconds.
Here, we are measuring only a special type of requests that are called watchlist requests. The new type of requests can be think of as LIST requests. Normal LIST requests are measure elsewhere. 

I hope that this answers your question.

Reply:
Yes that makes sense, thank you for the explanation.

I was thinking that for some reasons the LIST timeout was also applied to the watch list requests based on this note from the KEP:
> Note 2: Currently, there is a timeout for LIST requests of 60s. That means a slow reflector might fail synchronization as well and would have to re-establish the connection.